include/swift/RemoteInspection/TypeRefBuilder.h

```diff
@@ -486,6 +465,363 @@ class TypeRefBuilder {
+    /// added. Since we only add reflection infos, the ID can be its index.
+    /// We return a uint32_t since it's extremely unlikely we'll run out of
+    /// indexes.
+    uint32_t addReflectionInfo(ReflectionInfo I) {
+      ReflectionInfos.push_back(I);
+      auto InfoID = ReflectionInfos.size() - 1;
+      assert(InfoID <= UINT32_MAX && "ReflectionInfo ID overflow");
+      return InfoID;
+    }
+
+    const std::vector<ReflectionInfo> &getReflectionInfos() {
+      return ReflectionInfos;
+    }
+
+    /// Load unsubstituted field types for a nominal type.
+    RemoteRef<FieldDescriptor> getFieldTypeInfo(const TypeRef *TR);
+
+    /// Get the primitive type lowering for a builtin type.
+    RemoteRef<BuiltinTypeDescriptor> getBuiltinTypeInfo(const TypeRef *TR);
+
+    /// Get the raw capture descriptor for a remote capture descriptor
+    /// address.
+    RemoteRef<CaptureDescriptor> getCaptureDescriptor(uint64_t RemoteAddress);
+
+    /// Get the unsubstituted capture types for a closure context.
+    ClosureContextInfo getClosureContextInfo(RemoteRef<CaptureDescriptor> CD);
+
+    /// Get the multipayload enum projection information for a given TR
+    RemoteRef<MultiPayloadEnumDescriptor>
+    getMultiPayloadEnumInfo(const TypeRef *TR);
+
+    const TypeRef *lookupTypeWitness(const std::string &MangledTypeName,
+                                     const std::string &Member,
+                                     StringRef Protocol);
+
+    RemoteRef<char> readTypeRef(uint64_t remoteAddr);
+
+    template <typename Record, typename Field>
+    RemoteRef<char> readTypeRef(RemoteRef<Record> record, const Field &field) {
+      uint64_t remoteAddr = record.resolveRelativeFieldData(field);
+
+      return readTypeRef(remoteAddr);
+    }
+
+    llvm::Optional<std::string> normalizeReflectionName(RemoteRef<char> name);
+
+  private:
+    void populateFieldTypeInfoCacheWithReflectionAtIndex(size_t Index);
+    llvm::Optional<RemoteRef<FieldDescriptor>>
+    findFieldDescriptorAtIndex(size_t Index, const std::string &MangledName);
+
+    llvm::Optional<RemoteRef<FieldDescriptor>>
+    getFieldDescriptorFromExternalCache(const std::string &MangledName);
+
+    bool reflectionNameMatches(RemoteRef<char> reflectionName,
+                               StringRef searchName);
+
+    std::vector<ReflectionInfo> ReflectionInfos;
+
+    /// Indexes of Reflection Infos we've already processed.
+    llvm::DenseSet<size_t> ProcessedReflectionInfoIndexes;
+
+    /// Cache for field info lookups.
+    std::unordered_map<std::string, RemoteRef<FieldDescriptor>>
+        FieldTypeInfoCache;
+
+    /// Cache for normalized reflection name lookups.
+    std::unordered_map<uint64_t /* remote address */,
+                       llvm::Optional<std::string>>
+        NormalizedReflectionNameCache;
+
+    /// Cache for built-in type descriptor lookups.
+    std::unordered_map<std::string /* normalized name */,
```

Comment:
Not for this patch, but since we have LLVM here, this should probably be a StringMap for better efficiency. Same for all the other caches.